src/guardrails.rs

```diff
@@ -0,0 +1,369 @@
+                        ));
+                    }
+
+                    // Check if the calculus annotation is compatible with the encoding annotation
+                    if let Some(calculus) = &self.calculus {
+                        // If calculus is not allowed, return an error
+                        if !anno_ref.is_calculus_allowed(
+                            calculus,
+                            self.direction.ok_or(GuardrailError::AnnotationError(
+                                AnnotationError::NotInProcedure(s.span, *ident),
+                            ))?,
+                        ) {
+                            return Err(GuardrailError::UnsoundnessError(
+                                UnsoundnessError::CalculusEncodingMismatch(
+                                    self.direction.unwrap(),
+                                    s.span,
+                                    calculus.name,
+                                    anno_ref.name(),
+                                ),
+                            ));
+                        };
+                    }
+
+                    // Check if the annotation is a terminator annotation and set the flag
+                    if anno_ref.is_terminator() {
+                        self.terminator_annotation = Some(anno_ref.name());
+                    }
+                }
+            }
+            // If the statement is a block, increase the nesting level and walk the block
+            StmtKind::If(_, _, _)
+            | StmtKind::Angelic(_, _)
+            | StmtKind::Demonic(_, _)
+            | StmtKind::Seq(_) => {
+                if let Some(anno_name) = self.terminator_annotation {
+                    return Err(GuardrailError::UnsoundnessError(
+                        UnsoundnessError::NotTerminator(s.span, anno_name),
+                    ));
+                } else {
+                    self.nesting_level += 1;
+                    walk_stmt(self, s)?;
+                    self.nesting_level -= 1;
+                }
+            }
+            _ => {
+                // If there was a terminator annotation before, don't allow further statements
+                if let Some(anno_name) = self.terminator_annotation {
+                    return Err(GuardrailError::UnsoundnessError(
+                        UnsoundnessError::NotTerminator(s.span, anno_name),
+                    ));
+                } else {
+                    walk_stmt(self, s)?
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use crate::verify_test;
+
+    #[test]
+    fn test_wrong_annotation_usage() {
+        // this should produce an error
+        let source = r#"
+            proc main() -> () {
+                @invariant() // must be on a loop
+                var x: UInt
+            }
+        "#;
+        let res = verify_test(source).0.is_err(); // should fail
```

Comment:
Maybe check that the error message is actually the correct one? Or just match on a substring of the error message?